### DISCUSSION TITLE
HOWTO: Send intermediate messages back to the user while later chains are processing?

### DISCUSSION THREAD

**Discussion by dkindlund** [FlowiseAI/Flowise#1485] [Q&A] [answered]
When constructing an elaborate set of `LLM Chain` nodes, is there any way to provide the user with some sort of "intermediate progress" messages back through the chat history window -- while the user waits for the full results to be returned?

I'm thinking about a query that end-to-end might take 2+ minutes to complete.  Perhaps something like sending a message as an internal callback webhook via a `Custom Tool` to the front-end?

**Comment by Jaredude** [comment]
I don't think this is something that can be accomplished within a chain itself as an intermediate message. You could accomplish this via chat bot client code by streaming inserted text after the user has sent their message and then continue waiting and stream the text back from the server.

**Comment by dkindlund** [comment]
> You could accomplish this via chat bot client code by streaming inserted text after the user has sent their message and then continue waiting and stream the text back from the server.

Okay, this makes sense.  It would be nice if the default Flowise client code supported this feature, nonetheless.

**Reply by Jaredude** [reply]
After looking at the code for [FlowiseChatEmbed](https://github.com/FlowiseAI/FlowiseChatEmbed), I see it's using signals, which is great! I have an idea... I'll submit a PR to that project tomorrow.

**Reply by Jaredude** [reply, marked as answer]
Made a PR to the chatbot repo to support the ability to observe the signals inside the bot: https://github.com/FlowiseAI/FlowiseChatEmbed/pull/78

[Basic demo video](https://github-production-user-asset-6210df.s3.amazonaws.com/10228950/295727717-8fca6449-d1cf-4c0a-a5ba-274717ba40f1.mov)

**Reply by HenryHengZJ** [reply]
awesome @Jaredude !!

**Reply by dkindlund** [reply]
@Jaredude , this is really cool!  Thank you so much for this PR
